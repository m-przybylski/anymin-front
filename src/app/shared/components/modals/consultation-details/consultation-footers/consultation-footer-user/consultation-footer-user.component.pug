plat-consultation-footer
  //left part of the screen
  .consultation-footer__left-price
    plat-consultation-footer-price([grossPrice]='grossPrice')
  //middle part of the screen
  .consultation-footer__left-text([ngSwitch]='middlePanel')
    span.consultation-footer__text(*ngSwitchCase='middlePanelStatusTypes.freeMinute')
      span(plat-icon iconClass='information')
      p {{"CONSULTATION_DETAILS.FOOTER.FIRSTMINUTE" | translate}}
    span.consultation-footer__text(*ngSwitchCase='middlePanelStatusTypes.notAvailable')
      span(plat-icon iconClass='information')
      span
        p {{"CONSULTATION_DETAILS.FOOTER.NA_1" | translate}}
        p {{"CONSULTATION_DETAILS.FOOTER.NA_2" | translate}}
    span.consultation-footer__text(*ngSwitchCase='middlePanelStatusTypes.paymentCard')
      span(plat-icon iconClass='payment-card')
      p {{"CONSULTATION_DETAILS.FOOTER.CARD" |translate: {value: defaultPayment?.card | creditCard} }}
    span.consultation-footer__text(*ngSwitchCase='middlePanelStatusTypes.paymentAnyMind')
      span(plat-icon iconClass='wallet')
      p
        | {{"CONSULTATION_DETAILS.FOOTER.ACCOUNT" | translate }}
        strong  AnyMind
        | : {{ balance }} z≈Ç.
        span(*ngIf='duration') {{"CONSULTATION_DETAILS.FOOTER.TALKLAST" | translate:{duration: duration} }}
        a {{"CONSULTATION_DETAILS.FOOTER.TOPUP" | translate}}
  //right part of the screen
  .consultation-footer__right
    a(plat-large-icon-button, color='success', *ngIf="isExpertAvailable else offlineDesktop", (click)='onCall()')
      span(plat-icon iconClass='headphone')
      span {{ "CONSULTATION_DETAILS.FOOTER.CALL" | translate }}
    ng-template(#offlineDesktop)
      a(plat-large-icon-button color='primary', (click)='onNotifyUser()')
        span(plat-icon iconClass='bell')
        span {{ "CONSULTATION_DETAILS.FOOTER.NOTIFY" | translate }}
  //right panel on the screen on mobile
  .consultation-footer__right--mobile
    a(plat-large-icon-button color='success', *ngIf="isExpertAvailable else offlineMobile", (click)='onCall()')
      span(plat-icon iconClass='headphone')
      span {{ "CONSULTATION_DETAILS.FOOTER.CALL" | translate }}
    ng-template(#offlineMobile)
      a(plat-large-icon-button color='primary', (click)='onNotifyUser()')
        span(plat-icon iconClass='bell')
        span {{ "CONSULTATION_DETAILS.FOOTER.NOTIFY" | translate }}
