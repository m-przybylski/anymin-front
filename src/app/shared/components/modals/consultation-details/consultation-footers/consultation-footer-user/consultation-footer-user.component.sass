@import "~@anymind-ng/core/src/template/stylesheets/core/variables/colors"
@import "~@anymind-ng/core/src/template/stylesheets/core/helpers/mixins/mixins"
@import '../.../../../../../../../template/core/partials/grid'

.consultation-footer
  &__duration
    display: block

  &__left-price
    padding: 10px 0

  &__price
    &-money
      +h4
      font-weight: 300

    &-currency
      color: $c-half-white

  &__card
    padding: 0 10px 0 0
    display: block

    > div
      display: flex
      align-items: center

    span
      margin: 0 12px 0 0

    p
     +caption

  &__text
    display: inline-flex
    align-items: center
    padding: 0 10px 0 0

    > span
      margin: 0 12px 0 0

    p
      +caption

      strong
        font-weight: 700

      a
        +caption
        color: $c-white
        margin: 0
        font-weight: 700
        display: block

        +mdm
          display: inline

  &__right
    &-button
      margin: 0 0 0 16px

      &:nth-child(1)
        margin: 0

.button-large
  display: none

  +xsm
    display: inline-block

.button-small
  margin: 8px 0 0
  +xsm
    display: none
