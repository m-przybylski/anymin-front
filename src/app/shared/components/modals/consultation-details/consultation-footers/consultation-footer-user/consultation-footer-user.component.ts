import { Component, Inject, OnDestroy, OnInit } from '@angular/core';
import { Logger } from '@platform/core/logger';
import {
  IFooterOutput,
  CONSULTATION_FOOTER_DATA,
  IConsultationFooterData,
} from '@platform/shared/components/modals/consultation-details/consultation-footers/consultation-footer-helpers';
import { Observable, Subject, defer, of } from 'rxjs';
import { LoggerFactory, MoneyToAmount } from '@anymind-ng/core';
import { ConsultationDetailsActionsService } from '@platform/shared/components/modals/consultation-details/consultation-details-actions.service';
import { EmploymentWithExpertProfile, ExpertPresenceStatus } from '@anymind-ng/api';
import VatRateTypeEnum = EmploymentWithExpertProfile.VatRateTypeEnum;
import { Router } from '@angular/router';
import { RouterPaths } from '@platform/shared/routes/routes';
import { ExpertAvailabilityService } from '@platform/features/dashboard/components/expert-availability/expert-availablity.service';
import { startWith, map, takeUntil, mergeMap } from 'rxjs/operators';
import { CallStatusService } from '@platform/shared/components/modals/consultation-details/call-status.service';
import { Animations } from '@platform/shared/animations/animations';

export enum MiddlePanelStatusTypes {
  freeMinute,
  paymentCard,
  noPaymentMethod,
  notAvailable,
  freeConsultation,
  promoCodes,
}

@Component({
  templateUrl: 'consultation-footer-user.component.html',
  styleUrls: ['consultation-footer-user.component.sass'],
  animations: [Animations.fadeOut, Animations.fadeInWithDelay],
})
export class ConsultationFooterUserComponent extends Logger implements IFooterOutput, OnDestroy, OnInit {
  public get actionTaken$(): Observable<keyof ConsultationDetailsActionsService> {
    return this._actionTaken$.asObservable();
  }

  public get grossPrice(): string {
    return this.moneyPipe.transform(this.data.price) || this.moneyPipe.transform({ value: 0, currency: '' });
  }

  public get middlePanel(): Observable<MiddlePanelStatusTypes> {
    return this.data$.pipe(
      map(data => {
        if (data.userId === undefined) {
          return MiddlePanelStatusTypes.freeMinute;
        }
        if (!data.isExpertAvailable) {
          return MiddlePanelStatusTypes.notAvailable;
        }
        if (this.checkIsFreeConsultation(data)) {
          return MiddlePanelStatusTypes.freeConsultation;
        }
        if (data.defaultPaymentMethod.promoCodeId !== undefined) {
          return MiddlePanelStatusTypes.promoCodes;
        }

        if (data.defaultPaymentMethod.creditCardId === undefined) {
          return MiddlePanelStatusTypes.noPaymentMethod;
        }

        return MiddlePanelStatusTypes.paymentCard;
      }),
    );
  }
  public middlePanelStatusTypes = MiddlePanelStatusTypes;
  public isExpertAvailable$: Observable<boolean>;
  public isCallButtonLoading = false;
  public isNotifyRequestSended = false;

  public get card(): string {
    const currentCreditCard = this.data.creditCards.find(
      creditCard => creditCard.id === this.data.defaultPaymentMethod.creditCardId,
    );

    return currentCreditCard !== undefined ? `${currentCreditCard.cardType}${currentCreditCard.maskedNumber}` : '';
  }

  public get invoiceTrKey(): string {
    switch (this.data.vatRateType) {
      case VatRateTypeEnum.COMPANY0:
        return 'CONSULTATION_DETAILS.FOOTER.INVOICE_0';
      case VatRateTypeEnum.COMPANY23:
        return 'CONSULTATION_DETAILS.FOOTER.INVOICE_23';
      case VatRateTypeEnum.NATURALPERSON:
        return 'CONSULTATION_DETAILS.FOOTER.NO_INVOICE';
      default:
        this.loggerService.error('Unknown vat rate state');

        return '';
    }
  }

  private _actionTaken$ = new Subject<keyof ConsultationDetailsActionsService>();
  private moneyPipe = new MoneyToAmount(this.loggerService);
  private ngDestroy$ = new Subject<void>();
  private data$: Observable<IConsultationFooterData>;

  constructor(
    @Inject(CONSULTATION_FOOTER_DATA) private data: IConsultationFooterData,
    private expertAvailabilityService: ExpertAvailabilityService,
    private router: Router,
    private callStatusService: CallStatusService,
    loggerFactory: LoggerFactory,
  ) {
    super(loggerFactory.createLoggerService('ConsultationFooterUserComponent'));
  }

  public ngOnInit(): void {
    /** there will be one element if user have selected expert */
    this.isExpertAvailable$ = defer(() => {
      if (this.data.selectedExpertId !== undefined) {
        return this.expertAvailabilityService.getExpertPresence(this.data.selectedExpertId).pipe(
          map(presence => presence === ExpertPresenceStatus.StatusEnum.Available),
          startWith(this.data.isExpertAvailable),
        );
      }

      return of(this.data.isExpertAvailable);
    });

    this.data$ = of(this.data).pipe(
      mergeMap(data =>
        this.isExpertAvailable$.pipe(
          map(isExpertAvailable => ({
            ...data,
            isExpertAvailable,
          })),
        ),
      ),
    );

    this.callStatusService.callStatus$.pipe(takeUntil(this.ngDestroy$)).subscribe((isCallStarted: boolean) => {
      if (isCallStarted) {
        this.isCallButtonLoading = false;
      }
    });
  }

  public ngOnDestroy(): void {
    this._actionTaken$.complete();
    this.ngDestroy$.next();
    this.ngDestroy$.complete();
  }

  public onCall(): void {
    if (!this.isCallButtonLoading) {
      this.isCallButtonLoading = true;
      this._actionTaken$.next('makeCall');
    }
  }

  public onNotifyUser(): void {
    if (!this.isNotifyRequestSended) {
      this.isNotifyRequestSended = true;
      this._actionTaken$.next('notifyUser');
    }
  }

  public redirectToPayments(): void {
    this.router.navigate([RouterPaths.dashboard.user.payments.asPath]);
    this._actionTaken$.complete();
  }

  private checkIsFreeConsultation(data: IConsultationFooterData): boolean {
    return data.price !== undefined && data.price.value === 0;
  }
}
