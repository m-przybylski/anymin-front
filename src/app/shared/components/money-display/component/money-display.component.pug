| {{ sign }} {{ money.amount | moneyDivider | currency: money.currency:'symbol-narrow' }}
