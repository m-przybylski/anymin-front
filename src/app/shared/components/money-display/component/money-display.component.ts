import { Component, Input } from '@angular/core';
import { MoneyDto } from '@anymind-ng/api';

@Component({
  selector: 'plat-money-display',
  templateUrl: './money-display.component.html',
})
export class MoneyDisplayComponent {
  @Input()
  public set money(value: MoneyDto) {
    this._money = value;
  }

  public get money(): MoneyDto {
    return this._money || { value: 0, currency: 'PLN' };
  }

  @Input()
  public set sign(value: string) {
    this._sign = value;
  }

  public get sign(): string {
    if (this._sign && this.money.value !== 0) {
      return this._sign;
    }

    return '';
  }

  private _money: MoneyDto;
  private _sign: string;
}
