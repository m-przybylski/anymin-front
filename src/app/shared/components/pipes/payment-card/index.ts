import { NgModule } from '@angular/core';

import { CreditCardPipe } from './payment-card.pipe';

@NgModule({
  exports: [CreditCardPipe],
  declarations: [CreditCardPipe],
})
export class PaymentCardModule {}
