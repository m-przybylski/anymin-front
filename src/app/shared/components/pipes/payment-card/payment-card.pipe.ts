import { Pipe, PipeTransform } from '@angular/core';
import { DefaultCreditCard } from '@anymind-ng/api';

@Pipe({
  name: 'creditCard',
})
export class CreditCardPipe implements PipeTransform {
  public transform(value: DefaultCreditCard | undefined): string {
    if (!(value && value.card)) {
      return '';
    }

    return `${value.card.cardType} ${value.card.maskedNumber}`;
  }
}
