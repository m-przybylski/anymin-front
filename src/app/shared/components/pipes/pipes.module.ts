import { NgModule } from '@angular/core';
import { PaymentCardModule } from './payment-card';

@NgModule({
  imports: [PaymentCardModule],
  exports: [PaymentCardModule],
})
export class PipesModule {}
