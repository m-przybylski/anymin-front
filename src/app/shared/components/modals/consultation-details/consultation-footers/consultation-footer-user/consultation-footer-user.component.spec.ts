import { ConsultationFooterUserComponent, MiddlePanelStatusTypes } from './consultation-footer-user.component';
import { TestBed, async } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA, Pipe, PipeTransform } from '@angular/core';
import { PaymentCardModule } from '@platform/shared/components/pipes/payment-card';
import {
  CONSULTATION_FOOTER_DATA,
  IConsultationFooterData,
} from '@platform/shared/components/modals/consultation-details/consultation-footers/consultation-footer-helpers';
import { provideMockFactoryLogger } from 'testing/testing';

@Pipe({
  name: 'translate',
})
class PipeName implements PipeTransform {
  // tslint:disable-next-line:typedef
  public transform = (_: any) => _;
}

describe('ConsultationFooterUserComponent', () => {
  let consultationFooterUserComponent: ConsultationFooterUserComponent;
  let data: IConsultationFooterData = {} as any;
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [PaymentCardModule],
      schemas: [NO_ERRORS_SCHEMA],
      declarations: [ConsultationFooterUserComponent, PipeName],
      providers: [
        {
          provide: CONSULTATION_FOOTER_DATA,
          useFactory: (): IConsultationFooterData => data,
        },
        provideMockFactoryLogger(),
      ],
    }).compileComponents();
  }));

  it('should be created', () => {
    const componentFixture = TestBed.createComponent(ConsultationFooterUserComponent);
    consultationFooterUserComponent = componentFixture.componentInstance;
    expect(consultationFooterUserComponent).toBeTruthy();
  });
  it('should return MiddlePanelStatusTypes.notAvailable when expert is not available ', () => {
    data = {
      ownerId: 'asdf',
      expertsIdList: [] as any,
      isExpertAvailable: false,
      isFreelance: false,
      userId: '1234',
      price: { value: 100, currency: 'PLN' },
      defaultPaymentMethod: {
        creditCardId: 'id',
      },
      creditCards: [],
    };
    const componentFixture = TestBed.createComponent(ConsultationFooterUserComponent);
    consultationFooterUserComponent = componentFixture.componentInstance;

    expect(consultationFooterUserComponent.middlePanel).toEqual(MiddlePanelStatusTypes.notAvailable);
  });
  it('should return MiddlePanelStatusTypes.freeMinute when user is not logged', () => {
    data = {
      ownerId: 'asdf',
      expertsIdList: [] as any,
      isExpertAvailable: true,
      isFreelance: false,
      price: { value: 123, currency: 'PLN' },
      defaultPaymentMethod: {
        creditCardId: 'id',
      },
      creditCards: [],
    };
    const componentFixture = TestBed.createComponent(ConsultationFooterUserComponent);
    consultationFooterUserComponent = componentFixture.componentInstance;

    expect(consultationFooterUserComponent.middlePanel).toEqual(MiddlePanelStatusTypes.freeMinute);
  });
  it('should return MiddlePanelStatusTypes.freeMinute when user is not logged', () => {
    data = {
      ownerId: 'asdf',
      expertsIdList: ['asdf'] as any,
      isExpertAvailable: true,
      isFreelance: false,
      price: { value: 123, currency: 'PLN' },
      defaultPaymentMethod: {
        creditCardId: 'id',
      },
      creditCards: [],
    };
    const componentFixture = TestBed.createComponent(ConsultationFooterUserComponent);
    consultationFooterUserComponent = componentFixture.componentInstance;

    expect(consultationFooterUserComponent.middlePanel).toEqual(MiddlePanelStatusTypes.freeMinute);
  });

  it('should return MiddlePanelStatusTypes.paymentCard when user has card assigned', () => {
    data = {
      ownerId: 'asdf',
      expertsIdList: [] as any,
      isExpertAvailable: true,
      isFreelance: false,
      userId: '123',
      price: { value: 123, currency: 'PLN' },
      defaultPaymentMethod: {
        creditCardId: 'id',
      },
      creditCards: [],
    };
    const componentFixture = TestBed.createComponent(ConsultationFooterUserComponent);
    consultationFooterUserComponent = componentFixture.componentInstance;

    expect(consultationFooterUserComponent.middlePanel).toEqual(MiddlePanelStatusTypes.paymentCard);
  });

  it('should return false isExpertAvailable if userId is not provided', () => {
    data = {
      ownerId: 'asdf',
      expertsIdList: [] as any,
      isExpertAvailable: true,
      isFreelance: false,
      userId: undefined,
      price: { value: 100, currency: 'PLN' },
      defaultPaymentMethod: {},
      creditCards: [],
    };
    const componentFixture = TestBed.createComponent(ConsultationFooterUserComponent);
    consultationFooterUserComponent = componentFixture.componentInstance;

    expect(consultationFooterUserComponent.isExpertAvailable).toEqual(true);
  });
});
