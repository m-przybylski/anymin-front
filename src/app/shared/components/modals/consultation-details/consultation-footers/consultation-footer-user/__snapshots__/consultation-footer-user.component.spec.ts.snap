// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsultationFooterUserComponent should return true isExpertAvailable if userId is not provided 1`] = `
<ng-component
  _actionTaken$={[Function Subject]}
  callStatusService={[Function Object]}
  data={[Function Object]}
  data$={[Function Observable]}
  expertAvailabilityService={[Function Object]}
  isCallButtonLoading="false"
  isExpertAvailable$={[Function Observable]}
  isNotifyRequestSended="false"
  loggerService={[Function Object]}
  middlePanelStatusTypes={[Function Object]}
  moneyPipe={[Function MoneyToAmount]}
  ngDestroy$={[Function Subject]}
  router={[Function Object]}
>
  <plat-consultation-footer>
    <div
      class="consultation-footer__left-price"
    >
      <plat-consultation-footer-price />
    </div>
    <div
      class="consultation-footer__left-text"
      ng-reflect-ng-switch="0"
    >
      
      <span
        class="consultation-footer__text"
      >
        <span
          iconclass="information"
          plat-icon=""
        />
        <p>
          CONSULTATION_DETAILS.FOOTER.FIRSTMINUTE
        </p>
      </span>
      
      
      
      
      
    </div>
    <div
      class="consultation-footer__right"
    >
      
      <a
        color="success"
        plat-large-icon-button=""
      >
        <span
          iconclass="headphone"
          plat-icon=""
        />
        
        <span>
          CONSULTATION_DETAILS.FOOTER.CALL
        </span>
        
      </a>
      
    </div>
    <div
      class="consultation-footer__right--mobile"
    >
      <div
        class="button-large"
      >
        
        <a
          color="success"
          plat-large-icon-button=""
        >
          
          <span
            iconclass="headphone"
            plat-icon=""
          />
          
          <span>
            CONSULTATION_DETAILS.FOOTER.CALL
          </span>
          
        </a>
      </div>
      <div
        class="button-small"
      >
        
        <a
          color="success"
          plat-flat-fab=""
        >
          
          <span
            iconclass="headphone"
            plat-icon=""
          />
          
        </a>
      </div>
      
      
    </div>
  </plat-consultation-footer>
</ng-component>
`;
